csrc/optimization/alias_analysis.cpp

```diff
@@ -0,0 +1,96 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <unordered_map>
+#include <vector>
+
+#include <fusion.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+#include <optimization/alias_analysis.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+// Returns whether the input TensorView is allocated contiguously.
+bool isContiguous(const TensorView& tv) {
+  const std::vector<IterDomain*>& allocation_domain =
+      tv.getMaybeAllocationDomain();
+  for (size_t i = 0; i < allocation_domain.size(); i++) {
+    // We skip std::nullopt contiguity. It represents a broadcast or reduction
+    // dimension, which is of size 1 and always contiguous.
```

Comment:
Would we need to check that the size is indeed 1 to make this assumption?

I'm thinking of, say, a grouped convolution, which might reduce a 16-channel tensor to a 4-channel tensor. Might that 4-channel tensor still have an optional/currently-unresolved contiguity?

Reply:
> Would we need to check that the size is indeed 1 to make this assumption?

The reduction dimension holds the original extent, so technically it's not one but conceptually one for reasoning about contiguity... (I believe we were in the same knowledge share by @naoyam where I wondered the reason behind this design 😄 )

> I'm thinking of, say, a grouped convolution, which might reduce a 16-channel tensor to a 4-channel tensor. Might that 4-channel tensor still have an optional/currently-unresolved contiguity?

AFAIU, in this case, the fusion definition will rfactor the dimension of size 16 to 4x4 and reduce one of the 4s to 1. No?

Reply:
> I believe we were in the same knowledge share by @naoyam where I wondered the reason behind this design 😄 

Ahh, right, oops; yeah, I remember you asking that.

>> I'm thinking of, say, a grouped convolution, which might reduce a 16-channel tensor to a 4-channel tensor. Might that 4-channel tensor still have an optional/currently-unresolved contiguity?
>
> AFAIU, in this case, the fusion definition will rfactor the dimension of size 16 to 4x4 and reduce one of the 4s to 1. No?

Would it? I would imagine it would just leave it as a 4x4 in such a case, no?

Perhaps a better example is a 35-channel tensor that was grouped by 5 to get a 7-channel tensor. Unless the strides were forced on either tensor, as I understand it nvFuser would be free to choose a stride of 64 on the input side and 8 on the output side, instead of the densely-packed 35 and 7, respectively. For the smaller tensor this is a more reasonable choice, as we waste one measly element and in turn can get nicely-coalesced loads.

Then there's truly-dynamic tensors, where we can't even know at compile time if they're contiguous. This would be the group-by-5 example above, but with an unknown number of input channels.

As I understand it (and I could _easily_ be wrong), `nullopt` in the above case would mean either "nvFuser has not decided on a stride" or "nvFuser _can't_ decide on a stride (due to dynamism)". In such a situation we'd need to be conservative and assume non-contiguity, no?
(perhaps it would've made more sense for me to have put my question on the `if` from line 29, instead; oops, sorry!)

Reply:
Ha, I might have mislead you here @wujingyue 

Conceptually, `nullopt` should be indicating a broadcast stride. If that comes from either a `size-1` broadcast, or a reduction (which gives us an diminished dimension), it's safe to skip those and consider them as contiguous here.

In the meantime, if we have expanded dimensions, where we do require a stride == 0, I believe we also mark contiguity as `nullopt`? so for those cases, we cannot naively consider those as contiguous any more.

i.e. `at::randn(4, 5).unsqueeze(-1).expand((4, 5, 6)).reshape(40, 3)`

After the expand, the last dimension is a stride-0 dimension and we cannot split that dimension any more.

Reply:
Ouch, that hurts... Let me see what I can do. I need to detect dimensions that are an expanded broadcast.